### DISCUSSION TITLE
Add more inputs to logic/math components

### DISCUSSION THREAD

**Discussion by FalcoGer** [FakeFishGames/Barotrauma#9940] [Ideas and suggestions]
Often you want to add more than two numbers, or get the result of a chain of OR components. Right now you have to make long chains because all arithmetic and logic components are binary. I propose that all such components just get 4 inputs.

Calculations could be as follows:

`_` means a null signal (ie no signal)

For example:

 A | B | C | D | OR | AND | + | - | Notes
---|---|---|---|----|-----|---|---|------
 _ | _ | _ | _ | _  | _   | _ | _ | no inputs mean no outputs
 5 | _ | _ | _ | 1  | 1   | 5 | 5 | any value other than 0 is True for logic components. the sum of all values in the array [5] is 5.
 _ | 3 | _ | _ | 1  | 1   | 3 | 3 | first number present is subtrahend, division is the same. If you want -3, then you have to supply 0 in A.
 5 | 3 | _ | 0 | 1  | 0   | 8 | 2 | true and true and false is false, null signals are skipped

Obviously greater than and not wouldn't work like that and need to have 2 and 1 components respectively. Equals however is fine with 4 components, you just have to check if all of them are equal or not. With one input it should always be true.

Alternatively have a variable number of inputs, up to a limit of course. In that case if any input is null, the output is null.

**Comment by mygamingaccount** [comment]
Every logic component already has 11 inputs.

Join #baro-sub-building to learn more about how you can make use of this!

**Reply by FalcoGer** [reply]
Wires have this weird thing where the last signal "wins", except for certain components such as or and and, where actually all the signals are or-ed together on the same input pin.
This is inconsistent and illogical. Applying any logic or electrical engineering logic to the circuit world of this game makes little sense. Because of this, I think just adding more inputs would serve everybody better by removing confusing constructs like the one you describe.
Aside from that it also doesn't solve most of the problems I originally proposed.

Just dumping in 10 number signals into an add component certainly doesn't add them all together.

**Reply by mygamingaccount** [reply]
There's no inconsistency, when everything works as advertised.
Take OR for example: what is an OR component?
![image](https://user-images.githubusercontent.com/68995233/193370540-a61cb6c8-9639-40a7-a9dc-ba4e4750f9fe.png)
The OR component "sends a signal if either of the inputs receives a signal"
It doesn't matter how many wires connect to either inputs, if any of the two inputs receives a nonzero signal, that pin is active, then it will send its output.

AND component?
![image](https://user-images.githubusercontent.com/68995233/193370638-0f96e316-401e-4485-8eda-7bdd3d6071e4.png)
Sends a signal when **both** inputs receive a signal. So should it matter how many wires there are on the inputs, and which wires are active?

Output value is set by a signal sent to set_output. Again, no inconsistency there. Any signal will set the output, and if multiple components send to it, guess which will set the output? The one that sent last, overwriting every previous value.

You are assuming that everything should work as CMOS gates. As an engineer, you shouldn't assume, and instead read. Always read the documentation, read datasheets of the parts you work with.

> Just dumping in 10 number signals into an add component certainly doesn't add them all together.

Let me commend you for this excellent suggestion of how the adder should work.

**Reply by mygamingaccount** [reply]
I forgot to mention, you can have more than 11 inputs if you hook the inputs to junction box signal pins that you can use to combine signals from more components than you'll ever need. The relationship between the signals is OR, and can't be AND, but hey, they made computers from nothing but NOR gates that put men on the Moon, maybe it's just appropriate to use them on Europa.
https://en.wikipedia.org/wiki/Apollo_Guidance_Computer

**Reply by FalcoGer** [reply]
You miss the point. From an electrical engineering point, hooking multiple wires to the same terminal makes no sense. A junction box is huge and way overkill for any kind of logic circuit. And I said it's inconsistent and illogical, not it doesn't work as described.

Take ADD for example. It doesn't work AT ALL like Or. Take And for example, it doesn't work at all like Or. Or takes all the inputs and Ors them. And doesn't and the inputs, it ors them on the same terminal.

I say, one active input at a time only, more terminals. That makes it easier to understand, more logical, more consistent with real world circuits and it also shrinks circuit sizes in general. I don't see any downsides here. What are you arguing for?

The rest is an implementation issue. I think simply picking the number of input terminals in the object properties, up to a reasonable maximum, is probably the easiest. Of course some components are strictly binary and make no sense to have 3 or more inputs, such as `>`. Others like '=' make sense to have multiple inputs.

**Reply by mygamingaccount** [reply]
As an electrical engineer you hook up multiple wires to the same inputs very often. You're just not allowed to drive them at the same time. Open Collector logic lets you use a pull-up resistor and then anything can drive the input to GND. You should see a familiar pattern: this is how Barotrauma does it, just the other way, you drive a pin high instead of low.

> Take ADD for example. It doesn't work AT ALL like Or. Take And for example, it doesn't work at all like Or. Or takes all the inputs and Ors them. And doesn't and the inputs, it ors them on the same terminal.

It _shouldn't_ work like that at all, they are entirely dissimilar, one does boolean logic, the other does algebra. The adder component is not an n-bit series adder, its input is not wires, it uses floating point signals, so the distinction between it and electrical engineering practices ends there. you wouldn't want to OR the bits of the numbers now, would you?


Here's my point: if you change something, change it for the better. Your suggestion isn't better or worse, it's different. It's bad in a different way. Then why change anything? There's room for improvement with the wiring, and bigger things to think about. The whole system is a mess when you consider how timing and propagation delay affects circuits. That's what has to be fixed.

**Reply by FalcoGer** [reply]
From a gameplay perspective, they should work the same. Otherwise you end up with an inconsistent mess. Which is exactly what we got.

I might want to or the bits of the numbers. Because 0 is 0 and non 0 is 1, which is the behavior of logic operations on numbers in a lot of programming languages. I might also want bitwise operations on numbers. But that's besides the point.

I say the system should be consistent. And it isn't.

**Reply by Saevon** [reply]
Why are we comparing electrical engineering here?  clearly "wiring" in this game is NOT electronics, but signal logic. Its not intended for some low-level electronics folk, but for more general (not even a programming) crowd, as it is in most games like this.  The programming/electronics crowd just likes it a lot too

So the UI should be consistent with how all components work, they should work as expected. Do all components wait for ALL inputs?  and then wait for a "continuous mode" toggle if it shouldn't wait.

Like if you just want an "OR" right now to work differently (People should just wire multiple things to one input up to 11) then you can ALREADY do that with ANY WIRE TERMINAL. Then you want a "transform" (set output / false output) not an OR

**Comment by Saevon** [comment]
AND component:
- 5 inputs
- Sends an true output if all WIRED inputs are truthy (if nothing is wired it sends true)
- Waits till all WIRED connections have an input (null wired inputs means null output), unless "continuous mode" is on

OR component
- 5 inputs
- Sends an true output if any WIRED inputs are truthy (if nothing is wired it sends true)
- Waits till all WIRED connections have an input (null wired inputs means null output), unless "continuous mode" is on
- You can still wire multiple wires to one input in "continuous mode" if you want more then 6 inputs at once


ADD component
- 5 inputs
- Adds all WIRED inputs
- "Continuous mode" means it ignored NO-SIGNAL from any WIRED terminals

SUB component
- 1 input: `value_in`
- 4 inputs:  `sub_in_1`,,, `sub_in_5`
- Subtracts all WIRED inputs
- If `value_in` is not wired, then its just subtracting from 0
- "Continuous mode" means it ignored NO-SIGNAL from any WIRED terminals, treating them as unwired.

MULT component
- 5 inputs
- Multiplies all WIRED inputs
- "continuous mode" means all WIRED NO-SIGNAL inputs are ignored and all others are multiplied

CONCAT component
- 5 inputs
- joins all WIRED inputs
- "continous mode" treats all WIRED NO-SIGNAL inputs as UNWIRED



Setting this precedent means the following component becomes doable too!

MIN/MAX component:
- 5 inputs
- Sends the smallest/largest WIRED value given
- "continuous mode" constantly sends the smallest value, even if some WIRED values are NULL/NO-SIGNAL
  * sends NULL if all WIRED values are currently NULL/NO-SIGNAL

**Comment by dossy** [comment]
My wishlist improvement:

Change the Greater Component such that if `SIGNAL_IN_1` is greater than any of the values seen on `SIGNAL_IN_2` then `SIGNAL_OUT` will send the `Output` value, otherwise it sends the `False output` value.

Similarly, we need a Min Component and a Max Component that looks like the new Input Selector Component with 10 signal inputs and it outputs the minimum or maximum value of all non-null input signal values, corresponding to the component type (Min Component vs. Max Component, respectively).

The introduction of these two new components (Min Component, Max Component) would make changing the Greater Component to behave the way I described above unnecessary, and would open up a LOT of flexibility and reduce the currently expensive "chaining of components to build a circuit" problem.
